apps/gdalalg_raster_rgb_to_palette.cpp

```diff
@@ -0,0 +1,323 @@
+        return false;
+    }
+
+    const double oneOverStep =
+        1.0 / ((m_colorMap.empty() ? 1 : 0) + (bNeedTmpFile ? 2 : 1));
+
+    GDALColorTable oCT;
+
+    bool bOK = true;
+    double dfLastProgress = 0;
+    std::unique_ptr<void, decltype(&GDALDestroyScaledProgress)> pScaledData(
+        nullptr, GDALDestroyScaledProgress);
+    if (m_colorMap.empty())
+    {
+        pScaledData.reset(GDALCreateScaledProgress(0, oneOverStep, pfnProgress,
+                                                   pProgressData));
+        dfLastProgress = oneOverStep;
+        bOK = (GDALComputeMedianCutPCT(
+                   mapBands[GCI_RedBand], mapBands[GCI_GreenBand],
+                   mapBands[GCI_BlueBand], nullptr, m_colorCount,
+                   GDALColorTable::ToHandle(&oCT),
+                   pScaledData ? GDALScaledProgress : nullptr,
+                   pScaledData.get()) == CE_None);
+    }
+    else
+    {
+        GDALDriverH hDriver;
+        if ((hDriver = GDALIdentifyDriver(m_colorMap.c_str(), nullptr)) !=
+                nullptr &&
+            // Palette .txt files may be misidentified by the XYZ driver
+            !EQUAL(GDALGetDescription(hDriver), "XYZ"))
+        {
+            auto poPaletteDS = std::unique_ptr<GDALDataset>(GDALDataset::Open(
+                m_colorMap.c_str(), GDAL_OF_RASTER | GDAL_OF_VERBOSE_ERROR,
+                nullptr, nullptr, nullptr));
+            bOK = poPaletteDS != nullptr && poPaletteDS->GetRasterCount() > 0;
+            if (bOK)
+            {
+                const auto poCT =
+                    poPaletteDS->GetRasterBand(1)->GetColorTable();
+                if (poCT)
+                {
+                    oCT = *poCT;
+                }
+                else
+                {
+                    bOK = false;
+                    ReportError(CE_Failure, CPLE_AppDefined,
+                                "Dataset '%s' does not contain a color table",
+                                m_colorMap.c_str());
+                }
+            }
+        }
+        else
+        {
+            auto poCT = GDALColorTable::LoadFromFile(m_colorMap.c_str());
+            bOK = poCT != nullptr;
+            if (bOK)
+            {
+                oCT = std::move(*(poCT.get()));
+            }
+        }
+    }
+
+    std::unique_ptr<GDALDataset> poTmpDS;
+    if (bOK)
+    {
+        const std::string osTmpFilename =
+            bNeedTmpFile ? m_outputDataset.GetName() + ".tmp.tif"
+                         : m_outputDataset.GetName();
+        CPLStringList aosTmpCO;
+        if (!bNeedTmpFile)
+            aosTmpCO = CPLStringList(m_creationOptions);
+        poTmpDS.reset(poTmpDriver->Create(
+            osTmpFilename.c_str(), poSrcDS->GetRasterXSize(),
+            poSrcDS->GetRasterYSize(), 1, GDT_Byte, aosTmpCO.List()));
+        bOK = poTmpDS != nullptr;
+        if (bOK && bNeedTmpFile)
+        {
+            VSIUnlink(osTmpFilename.c_str());
```

Comment:
Am I misreading this or you are deleting the temp file created at line 254?

Reply:
I guess I misinterpreted what bNeedTmpFile means, does it mean that it needs to be kept instead of unlinked?